app/models/user.rb

```diff
@@ -158,6 +159,51 @@ def self.authenticate(options)
     user
   end
 
+  aasm :column => :status, :no_direct_assignment => true do
+    state :pending, :initial => true
+    state :active
+    state :confirmed
+    state :suspended
+    state :deleted
+
+    # A normal account is active
+    event :activate do
+      transitions :from => :pending, :to => :active
+    end
+
+    # Used in test suite, not something that we would normally need to do.
+    event :deactivate do
+      transitions :from => :active, :to => :pending
+    end
+
+    # To confirm an account is used to override the spam scoring
+    event :confirm do
+      transitions :from => [:pending, :active, :suspended], :to => :confirmed
+    end
+
+    event :suspend do
+      transitions :from => [:pending, :active], :to => :suspended
+    end
+
+    # Mark the account as deleted but keep all data intact
+    event :hide do
+      transitions :from => [:pending, :active, :confirmed, :suspended], :to => :deleted
+    end
+
+    event :unhide do
+      transitions :from => [:deleted], :to => :active
+    end
```

Comment:
This signals to me that hiding is not really a state but a flag on the user. A bit like act_as_paranoid ([and similar](https://github.com/rubysherpas/paranoia)) did back in the day. 

Using a flag would clarify the `:unhide` event. Right now, a `pending` user that gets deleted would become `active` without ever running through the activation steps.

Another way to solve this would be to transition to `pending` within `unhide` and ask the user to re-take the activation steps.

---

However, I wonder what the use case for hiding an account would be. I had assumed that `remove_personal_data` was applied to all deleted users. Which IMO is desirable since we don't want personal info lying around on accounts that are not used and not visible, right? Could we maybe remove the `hide` usecase altogether?

Reply:
It's used for spam accounts, to close them without removing the evidence.